File: src/_deprecated/Detection.cpp
#include <thread>
#include <atomic>
#include <chrono>
#include <vector>
#include <iostream>
#include <windows.h>
#include "SharedMemory.h"
#include <opencv2/opencv.hpp>

const double THRESHOLD = 0.9;

cv::Mat tossImg = cv::imread("assets/1920x1080/normal/toss.png", cv::IMREAD_COLOR);
cv::Mat serveImg = cv::imread("assets/1920x1080/normal/serve.png", cv::IMREAD_COLOR);

static bool checkTemplatesLoaded()
{
    if (tossImg.empty() || serveImg.empty())
    {
        std::cerr << "One or both template images not found!" << std::endl;
        return false;
    }
    return true;
}

static HWND getRobloxWindow()
{
    HWND robloxHwnd = FindWindow(NULL, L"Roblox");
    HWND activeHwnd = GetForegroundWindow();
    if (robloxHwnd && activeHwnd == robloxHwnd)
    {
        return robloxHwnd;
    }
    return NULL;
}

static bool getRobloxWindowRect(int &left, int &top, int &width, int &height)
{
    HWND robloxHwnd = FindWindow(NULL, L"Roblox");
    if (!robloxHwnd)
    {
        std::cerr << "Roblox window not found!" << std::endl;
        return false;
    }
    RECT rect;
    if (!GetWindowRect(robloxHwnd, &rect))
    {
        std::cerr << "Failed to get window rect: " << GetLastError() << std::endl;
        return false;
    }
    left = rect.left;
    top = rect.top;
    width = rect.right - rect.left;
    height = rect.bottom - rect.top;
    return true;
}

static cv::Mat captureRegion(HWND hwnd, int x, int y, int width, int height)
{
    if (!hwnd)
        return cv::Mat();
    HDC hWindowDC = GetDC(hwnd);
    if (!hWindowDC)
        return cv::Mat();
    HDC hMemoryDC = CreateCompatibleDC(hWindowDC);
    HBITMAP hBitmap = CreateCompatibleBitmap(hWindowDC, width, height);
    SelectObject(hMemoryDC, hBitmap);
    BitBlt(hMemoryDC, 0, 0, width, height, hWindowDC, x, y, SRCCOPY);
    BITMAPINFOHEADER bi = {sizeof(BITMAPINFOHEADER), width, -height, 1, 32, BI_RGB};
    std::vector<unsigned char> buffer(width * height * 4);
    GetDIBits(hMemoryDC, hBitmap, 0, height, buffer.data(), (BITMAPINFO *)&bi, DIB_RGB_COLORS);
    cv::Mat img(height, width, CV_8UC4, buffer.data());
    cv::Mat bgr;
    cv::cvtColor(img, bgr, cv::COLOR_BGRA2BGR);
    DeleteObject(hBitmap);
    DeleteDC(hMemoryDC);
    ReleaseDC(hwnd, hWindowDC);
    return bgr;
}

static void serveLoop(SharedMemory &shm)
{
    while (true)
    {
        shm.setBool(ShmVariable::isInServeLoop, true);
        if (!shm.getBool(ShmVariable::isServing))
        {
            shm.setBool(ShmVariable::isInServeLoop, false);
            break;
        }
        HWND robloxHwnd = getRobloxWindow();
        int left, top, width, height;
        if (robloxHwnd && getRobloxWindowRect(left, top, width, height))
        {
            // cv::Mat screenshot = captureRegion(robloxHwnd, 854, 849, 193, 13);
            cv::Mat screenshot = captureRegion(robloxHwnd, 854, 849, 13, 13);
            // cv::Mat screenshot = captureRegion(robloxHwnd, left, top, width, height);
            // std::cout << left << " " << top << " " << width << " " << height << std::endl;
            if (!screenshot.empty())
            {
                cv::Mat result;
                cv::matchTemplate(screenshot, serveImg, result, cv::TM_CCOEFF_NORMED);
                double maxVal;
                cv::Point maxLoc;
                cv::minMaxLoc(result, nullptr, &maxVal, nullptr, &maxLoc);
                if (maxVal >= 0.9)
                {
                    // std::cout << "FOUND" << std::endl;
                    // std::cout << "Serve detected at: " << maxLoc.x << ", " << maxLoc.y << ", " << serveImg.cols << ", " << serveImg.rows << " " << maxVal << std::endl;
                    std::this_thread::sleep_for(std::chrono::milliseconds(500));
                    shm.setBool(ShmVariable::serveCommand, true);
                    shm.setBool(ShmVariable::isInServeLoop, false);
                    break;
                }
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
}

static void tossLoop(SharedMemory &shm)
{
    while (true)
    {
        HWND robloxHwnd = getRobloxWindow();
        int left, top, width, height;
        if (robloxHwnd && getRobloxWindowRect(left, top, width, height))
        {
            cv::Mat screenshot = captureRegion(robloxHwnd, 918, 1050, 74, 11);
            if (!screenshot.empty())
            {
                cv::Mat result;
                cv::matchTemplate(screenshot, tossImg, result, cv::TM_CCOEFF_NORMED);
                double maxVal;
                cv::Point maxLoc;
                cv::minMaxLoc(result, nullptr, &maxVal, nullptr, &maxLoc);
                if (maxVal >= THRESHOLD)
                {
                    shm.setBool(ShmVariable::isServing, true);
                    if (!shm.getBool(ShmVariable::isInServeLoop))
                    {
                        std::thread(serveLoop, std::ref(shm)).detach();
                    }
                }
                else
                {
                    shm.setBool(ShmVariable::isInServeLoop, false);
                    shm.setBool(ShmVariable::isServing, false);
                    shm.setBool(ShmVariable::isServe, false);
                    shm.setBool(ShmVariable::isToss, false);
                    shm.setBool(ShmVariable::serveCommand, false);
                }
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
}

int main()
{
    SharedMemory shm;
    if (!checkTemplatesLoaded())
        return 1;
    if (!shm.initialize())
        return 1;
    tossLoop(shm);
    return 0;
}

File: src/_deprecated/detection.py
import cv2
import main
import time
import win32gui
import pyautogui
import numpy as np
import pygetwindow as gw
import threading

def start():
    toss_loop()

tossImg = cv2.imread("assets/1920x1080/normal/toss.png", cv2.IMREAD_COLOR)
serveImg = cv2.imread("assets/1920x1080/normal/serve.png", cv2.IMREAD_COLOR)

if tossImg is None or serveImg is None:
    print("One or both template images not found!")
    exit()

threshold = 0.9

def get_roblox_window():
    roblox_hwnd = win32gui.FindWindow(None, "Roblox")
    active_hwnd = win32gui.GetForegroundWindow()
    if active_hwnd == roblox_hwnd:
        try:
            win = gw.getWindowsWithTitle("Roblox")[0]
            return win.left, win.top, win.width, win.height
        except Exception as e:
            print(f"Error getting window: {e}")
    return None

def toss_loop():
    while True:
        window_coords = get_roblox_window()
        if window_coords:
            left, top, width, height = window_coords
            region1 = (left, top, width, height)
            # region1 = (left + (width // 2) - 32, top + height - 24, 55, 13) # 1604x720
            # region1 = (674, 608, 50, 13) # 1404x630 
            # region1 = (677, 611, 44, 10) # 1404x630 new
            # region1 = (642, 744, 72, 10)  # 1366x768
            # region1 = (915, 1048, 81, 14) # 1920x1080
            region1 = (918, 1050, 74, 11) # 1920x1080 normal
            try:
                screenshot1 = pyautogui.screenshot(region=region1)
                screenshot_cv1 = cv2.cvtColor(np.array(screenshot1), cv2.COLOR_RGB2BGR)
                result1 = cv2.matchTemplate(screenshot_cv1, tossImg, cv2.TM_CCOEFF_NORMED)
                _, max_val1, _, max_loc = cv2.minMaxLoc(result1)
                if max_val1 >= threshold:
                    # print((max_loc[0], max_loc[1], tossImg.shape[1], tossImg.shape[0]), max_val1)
                    main.isServing = True
                    if not main.isInServeLoop:
                        threading.Thread(target=serve_loop, daemon=True).start()
                else:
                    main.isServing = False
                    main.isServe = False
                    main.isToss = False
            except Exception as e:
                print(f"Error in toss detection: {e}")
        time.sleep(0.2)

def serve_loop():
    while True:
        main.isInServeLoop = True
        window_coords = get_roblox_window()
        if not main.isServing:
            main.isInServeLoop = False
            break
        if window_coords and main.isServing and not main.isServe:
            left, top, width, height = window_coords
            region2 = (left, top, width, height)
            # region2 = (left + (width // 2) - 95, top + height - 156, 125, 55) # 1604x720
            # region2 = (615, 493, 18, 8) # 1404x630
            # region2 = (595, 602, 29, 11) # 1366x768
            # region2 = (853, 849, 27, 13) # 1920x1080
            region2 = (855, 850, 34, 12) # 1920x1080 normal
            try:
                screenshot2 = pyautogui.screenshot(region=region2)
                screenshot_cv2 = cv2.cvtColor(np.array(screenshot2), cv2.COLOR_RGB2BGR)
                result2 = cv2.matchTemplate(screenshot_cv2, serveImg, cv2.TM_CCOEFF_NORMED)
                _, max_val2, _, max_loc = cv2.minMaxLoc(result2)
                if max_val2 >= threshold:
                    # print((max_loc[0], max_loc[1], serveImg.shape[1], serveImg.shape[0]), max_val2)
                    time.sleep(0.48)
                    main.serve()
                    main.isInServeLoop = False
                    break
            except Exception as e:
                print(f"Error in serve detection: {e}")
        time.sleep(0.05)